src/xpcc/driver/pressure/bmp085_data.hpp

```diff
@@ -0,0 +1,191 @@
+
+#include <stdio.h>
+#include <cmath>
+
+#include <xpcc/debug/logger/logger.hpp>
+
+#ifndef XPCC_BMP085_DATA_HPP
+#define XPCC_BMP085_DATA_HPP
+
+#undef  XPCC_LOG_LEVEL
+#define XPCC_LOG_LEVEL xpcc::log::DISABLED
+
+// Forward declaration the test class
+class Bmp085Test;
+
+namespace xpcc
+{
+
+template < typename I2cMaster >
+class Bmp085;
+
+namespace bmp085data
+{
+
+/**
+ * Holds the calibration data from the sensor.
+ * Values are used for calculation of calibrated
+ * sensor values from raw sensor data
+ */
+struct ATTRIBUTE_PACKED
+Calibration
+{
+	int16_t  ac1;
+	int16_t  ac2;
+	int16_t  ac3;
+	uint16_t ac4;
+	uint16_t ac5;
+	uint16_t ac6;
+
+	int16_t  b1;
+	int16_t  b2;
+
+	int16_t  mb;
+	int16_t  mc;
+	int16_t  md;
+};
+
+enum class
+CalculationMode
+{
+	FixedPoint,
+	// SinglePrecision,
+	DoublePrecision,
+};
+
+template < CalculationMode >
+class ATTRIBUTE_PACKED
+DataBase
+{
+	template < typename I2cMaster >
+	friend class Bmp085;
+
+	// Grant unit test full access to private members.
+	friend class ::Bmp085Test;
+
+public:
+	/// DATA Access
+	inline Calibration &
+	getCalibration() {
+		return calibration;
+	}
+
+public:
+	/// Notify data class about changed buffer of raw temperature data.
+	void inline
+	rawTemperatureTouched() { meta &= ~TEMPERATURE_CALCULATED; }
+
+	/// Notify data class about changed buffer of raw pressure data.
+	void inline
+	rawPressureTouched() { meta &= ~PRESSURE_CALCULATED; }
+
+	/**
+	 * Get the calibrated temperature for the device in 0.1 degree Celsius
+	 *
+	 * If recalculation is necessary it is done on the fly.
+	 * No I2C transaction.
+	 */
+	int16_t inline
+	getTemperature();
+
+	void inline
+	getTemperature(int16_t &temp) { temp = getTemperature(); };
+
+	void inline
+	getTemperature(float &temp);
+
+	void inline
+	getTemperature(double &temp);
+
+	/**
+	 * Get the calibrated pressure from the device in Pascal.
+	 *
+	 * If recalculation is necessary it is done on the fly.
+	 * No I2C transaction.
+	 */
+	int32_t inline
+	getPressure();
+
+private:
+	/**
+	 * Use the calibration data read from the sensor to
+	 * calculate the calibrated temperature from the
+	 * raw data.
+	 * The result is stored in this struct for further
+	 * access.
+	 * Implemented differently for Fixed Point, Single and
+	 * Double Precision by template specialsation.
+	 */
+	void inline
+	calculateCalibratedTemperature();
+
+	/**
+	 * See calculateCalibratedTemperature()
+	 */
+	void inline
+	calculateCalibratedPressure();
+
+	// The order of these private variable is optimized for alignment of 4
+	Calibration calibration;
+
+private:
+	// Fixed Point
+	int16_t calibratedTemperature; // in 0.1 degree Celsius
+	int32_t calibratedPressure;    // in Pa
+
+	int32_t b5; // calculated in calculateCalibratedTemperature, needed for calculateCalibratedPressure
+
+private:
+	// Double Precision
+	// Prepared coefficients for temperature conversion in double precision
+	double c5, c6, mc, md;
```

Comment:
> Also note, that your `getTemperature()` function can now have different return types!

But why should it?

The aim of the encapsulation of the calculation is that the user code can be the same and only the driver differs in precision, speed and code size. I would add getters with references (`getTemperature(& int16_t)`, `getTemperature(& float)` and `getTemperature(& double)`)

BTW, I would suggest an xpcc::TemperatureFp data type which is `int16_t` in 0.01 degrees. That's between -160 and +160 degrees and is enough for all environmental sensors, including BME280.